### PR TITLE
[MM14489] Fix: In interactive dialogs, render autocomplete lists below the input field by default

### PR DESCRIPTION
**Pull request by larkox**
#### Summary

Fix: In interactive dialogs, render autocomplete lists below the input field by default
#### Ticket Link

https://mattermost.atlassian.net/browse/MM-14489
#### Related Pull Requests

None

#### Screenshots

None

### CODE REVIEW

**Comment by asaadmahmood** [comment]
@jasonblais @prapti the autocomplete is now opening at the bottom, but I think it is now always opening at the bottom. The expected should be, it should open at the bottom or at the top depending on where it can find most space.

Additionally, the dropdown is pushing content and producing a scroll, rather than opening on top of the content, which should not be the case imo.

Dec-29-2019 12-48-06

**Comment by jasonblais** [comment]
@asaadmahmood yes, that makes sense to me. Do we have any existing components/dialogs that does that, so @larkox can use it as an example and apply it here?

**Comment by asaadmahmood** [comment]
@jasonblais The more actions menu does that, it:
1. Opens on top of content.
2. Toggles if we click the button again.
3. Positions itself depending on the space available.

Screenshot 2019-12-30 at 8 33 37 AM

**Comment by jasonblais** [comment]
Great, thank you! @larkox let us know if the above is helpful for you in terms of next steps?

**Comment by larkox** [comment]
> @jasonblais @prapti the autocomplete is now opening at the bottom, but I think it is now always opening at the bottom. The expected should be, it should open at the bottom or at the top depending on where it can find most space.
> Dec-29-2019 12-48-06

Right now, if the logic is working, it should drop down by default unless it has no space below. In that case, it will drop up. The logic behind the more actions button is similar.

**Comment by jasonblais** [comment]
Thanks @larkox! Looks like currently the autocomplete opens outside of the dialog

image

File: components/suggestion/suggestion_box.jsx [hunk 1]
```diff
@@ -690,24 +700,26 @@ export default class SuggestionBox extends React.Component {
...
-                        onItemHover={this.setSelection}
-                        cleared={this.state.cleared}
-                        matchedPretext={this.state.matchedPretext}
-                        items={this.state.items}
-                        terms={this.state.terms}
-                        selection={this.state.selection}
-                        components={this.state.components}
-                        wrapperHeight={this.props.wrapperHeight}
-                    />
+                    <div style={{position: 'fixed', zIndex: 101, width: this.state.width}}>
```
Comment (hmhealey) [inline components/suggestion/suggestion_box.jsx]: This isn't going to work since it breaks the positioning of other autocompletes throughout the app (notably the ones in the post textbox and search box). To get around that, you could make another SuggestionList-like component for this case and use that instead.

If you do that, don't look at `SearchSuggestionList` when doing that though. It has `SearchSuggestionList extends` the base `SuggestionList` which causes some problems. You could instead make your new component just render a `SuggestionList` with a wrapper div around it instead.

**Comment by hanzei** [comment]
Hi @larkox,

just wanted to check if you have any questions about the provided feedback

**Comment by larkox** [comment]
@hanzei Sorry for the delay. I got a bit stuck with this PR. I will give it another try this week, and see if I can make it work!

**Comment by larkox** [comment]
Current solution should work, but is by far not the most desirable.

It is flaky, due to being really dependant on the DOM, and not performing due to too many calls to find nodes in the DOM in too many "willUpdateComponent".

I will try to figure out ways to improve this, but I would love if you could propose any improvement, as I have been very blocked in this issue, and even if I got it to work (I think) I feel I am starting to get tunnel vision.

Thanks!

**Comment by asaadmahmood** [comment]
A few issues I'm seeing.
1. The dropdown opens in weird location at times, try resizing your browser height and then try opening a dropdown.
Screenshot 2020-01-20 at 11 05 01 PM

2. Sometimes, when I hover over the options, it seems like various options are selected on hover.
Screenshot 2020-01-20 at 11 04 35 PM

**Comment by larkox** [comment]
@asaadmahmood Thank you for finding those issues. This fix is getting harder than I expected.

I will see what can I do with the first issue.

Regarding the second problem, I will double check, but should not be specifically related to this issue. If several options have the same ID, those options are marked as hover instead of just one. In theory, this should not happen in real scenarios.

**Comment by larkox** [comment]
Now the code should be less brittle and more efficient. @hmhealey it would be nice if you could do another review now to see other points of improvement.

@asaadmahmood Not sure why the test server is not launching, but current version should have corrected the problem you found.

File: components/suggestion/modal_suggestion_list.jsx [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    componentDidUpdate() {
+        if (!this.scrollRegistered) {
+            this.tryRegisterScroll();
+        }
+
+        if (this.getChildHeight() === 0) {
+            return;
+        }
+
+        const newInputBounds = this.updateInputBounds();
```
Comment (hmhealey) [inline components/suggestion/modal_suggestion_list.jsx]: Generally, we want code in `componentDidUpdate` to only trigger when there's some sort of change in the props or state since since updates can happen due to unrelated state changes. Can this be wrapped in something like `if (this.state.scroll !== prevState.scroll) {...}`?

File: components/suggestion/modal_suggestion_list.jsx [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        this.updateModalBounds();
+    }
+
+    getChildHeight = () => {
+        if (!this.container.current) {
+            return 0;
+        }
+
+        const container = ReactDOM.findDOMNode(this.container.current);
+        const listElement = container.querySelector('.suggestion-list__content');
```
Comment (hmhealey) [inline components/suggestion/modal_suggestion_list.jsx]: Instead of using a query selector, could we attach a ref to the `SuggestionList` and use its `getContent` method?

File: components/suggestion/modal_suggestion_list.jsx [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+        this.updateModalBounds();
+    }
+
+    getChildHeight = () => {
+        if (!this.container.current) {
+            return 0;
+        }
+
+        const container = ReactDOM.findDOMNode(this.container.current);
```
Comment (hmhealey) [inline components/suggestion/modal_suggestion_list.jsx]: You can actually remove the `ReactDOM.findDOMNode` here. If a ref is on a regular HTML element, you'll already get the DOM node.

File: components/suggestion/suggestion_box.jsx [hunk 5]
```diff
@@ -194,6 +195,14 @@ export default class SuggestionBox extends React.Component {
         EventEmitter.removeListener('mention_key_click', this.handleMentionKeyClick);
     }
 
+    calculateInputRect = () => {
+        if (this.inputRef.current) {
+            const rect = ReactDOM.findDOMNode(this.inputRef.current).getBoundingClientRect();
```
Comment (hmhealey) [inline components/suggestion/suggestion_box.jsx]: Instead of using `ReactDOM` here, you might be able to use `this.inputRef.current.getInput()`

File: components/autocomplete_selector.jsx [hunk 6]
```diff
@@ -129,7 +129,7 @@ export default class AutocompleteSelector extends React.PureComponent {
                     <SuggestionBox
                         placeholder={placeholder}
                         ref={this.setSuggestionRef}
-                        listComponent={SuggestionList}
+                        listComponent={ModalSuggestionList}
```
Comment (hmhealey) [inline components/autocomplete_selector.jsx]: Is there anywhere else that we use the `AutocompleteSelector` that's outside of a modal?

File: components/suggestion/modal_suggestion_list.jsx [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    componentWillUnmount() {
+        if (this.container.current && this.scrollRegistered) {
+            const modalBodyContainer = getClosestParent(ReactDOM.findDOMNode(this.container.current), '.modal-body');
+            modalBodyContainer.removeEventListener('scroll', this.onModalScroll);
+        }
+        window.removeEventListener('resize', this.updateModalBounds);
+    }
+
+    componentDidUpdate() {
+        if (!this.scrollRegistered) {
```
[the remark this answers is not in the document]
Comment (larkox) [reply components/suggestion/modal_suggestion_list.jsx]: I am not sure. It gets registered on mount as long as the reference has an element. There should not be any instance where the reference does not includes an element, but I was safe-checking that. If you know the reference will always include an element, I can remove these checks.

File: components/suggestion/modal_suggestion_list.jsx (hunk 2 above)
Comment (larkox) [reply components/suggestion/modal_suggestion_list.jsx]: Using only the scroll would not be enough. Resizing the screen (and therefore changing the modal bounds) may not change the scroll, and that would not work. Also, a change in the child size (that means the selector is opened) may not change the scroll, which will also not work.

Nevertheless, I like the idea of only doing these changes whenever they are needed. I will upload an approach on the next commit.

File: components/autocomplete_selector.jsx (hunk 6 above)
Comment (larkox) [reply components/autocomplete_selector.jsx]: Yes, it is. It is used on ActionMenu component, which is used on the MessageAttachment component. I did not see that. I will solve this on the next commit.

File: components/suggestion/modal_suggestion_list.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/suggestion/modal_suggestion_list.jsx]: The div with the ref is always returned by `render`, so the ref should always be set before you get to `componentDidMount`. That means we should be able to remove this from `componentDidUpdate` and remove the "try" part of `tryRegisterScroll`

**Comment by hmhealey** [comment]
Thanks for making those changes, @larkox. I followed up on the one comment above if you could take a look at it.

File: components/suggestion/modal_suggestion_list.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (larkox) [reply components/suggestion/modal_suggestion_list.jsx]: Done

File: components/suggestion/suggestion_box.jsx [hunk 8]
```diff
@@ -693,24 +701,28 @@ export default class SuggestionBox extends React.Component {
...
+                            preventClose={this.preventSuggestionListClose}
+                            onItemHover={this.setSelection}
+                            cleared={this.state.cleared}
+                            matchedPretext={this.state.matchedPretext}
+                            items={this.state.items}
+                            terms={this.state.terms}
+                            selection={this.state.selection}
+                            components={this.state.components}
+                            wrapperHeight={this.props.wrapperHeight}
+                            calculateInputRect={this.calculateInputRect}
```
Comment (ali-farooq0) [inline components/suggestion/suggestion_box.jsx]: rather than `suggestion_box` calculating the react size and returning that, it'd be better design wise if you either pass in the input and let it calculate the bounds or passed the sizes as a prop, rather than a function and let it be reactive rather than invoking a function to get the sizes.
e.g. a component like this https://github.com/souporserious/react-measure could help make this reactive

File: components/suggestion/modal_suggestion_list.jsx [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        if (this.state.scroll !== e.target.scrollTop &&
+            this.latestHeight !== 0) {
+            this.setState({scroll: e.target.scrollTop});
+        }
+    }
+
+    componentDidMount() {
+        if (this.container.current) {
+            const modalBodyContainer = getClosestParent(this.container.current, '.modal-body');
+            modalBodyContainer.addEventListener('scroll', this.onModalScroll);
```
Comment (ali-farooq0) [inline components/suggestion/modal_suggestion_list.jsx]: this seems _very_ brittle. Would you be able to figure out where this component is if you received x,y,width of the input component? if so i would recommend using that this scroll change events.

File: components/suggestion/modal_suggestion_list.jsx [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+            this.setState({modalBounds: {top: modalBounds.top, bottom: modalBounds.bottom}});
+        }
+    }
+
+    render() {
+        const {
+            ...props
+        } = this.props;
+
+        Reflect.deleteProperty(props, 'onLoseVisibility');
```
[the remark this answers is not in the document]
Comment (larkox) [reply components/suggestion/modal_suggestion_list.jsx]: The problem in doing it your way is that we would declare two variables that will never be used, and if I am not mistaken, will generate some lint errors.

File: components/autocomplete_selector.jsx [hunk 11]
```diff
@@ -20,13 +20,15 @@ export default class AutocompleteSelector extends React.PureComponent {
         footer: PropTypes.node,
         disabled: PropTypes.bool,
         toggleFocus: PropTypes.func,
+        listComponent: PropTypes.func,
```
[the remark this answers is not in the document]
Comment (larkox) [reply components/autocomplete_selector.jsx]: I followed the example on SuggestionBox. My guess is that it is not a node by itself, but the constructor of a node of that class. Makes sense?

File: components/suggestion/modal_suggestion_list.jsx [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+        let position = {};
+        if (this.state.location === 'top') {
+            position = {bottom: this.state.modalBounds.bottom - this.state.inputBounds.top};
+        } else {
+            position = {top: this.state.inputBounds.bottom - this.state.modalBounds.top};
+        }
+
+        return (
+            <div
```
[the remark this answers is not in the document]
Comment (larkox) [reply components/suggestion/modal_suggestion_list.jsx]: After some checks, it seems that Overlay will position fine, but will not react to the scroll. Also, we would still need to control the width of the component.

File: components/suggestion/modal_suggestion_list.jsx (hunk 9 above)
Comment (larkox) [reply components/suggestion/modal_suggestion_list.jsx]: I can figure out the position of the component if I have x, y, and width, yes. The problem is that whenever the modal scrolls, those values would change. if we could pass x, y and width as props instead of the calculateInputRect function, there would be no need for checking this scroll.

Also, this class name is not something we can control. It comes from react-bootstrap's Modal. Which makes it a bit more robust (we cannot change that class name unless we use something different as a modal) and brittle at the same time (any change on react-bootstrap's Modal that change that will break the application).

Nevertheless, I think it would be very difficult to pass these values through props. Since it is more related to the previous comment, I will explain there.

File: components/suggestion/modal_suggestion_list.jsx [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+import SuggestionList from 'components/suggestion/suggestion_list.jsx';
+import {getClosestParent} from 'utils/utils.jsx';
+
+export default class ModalSuggestionList extends React.PureComponent {
+    static propTypes = {
+        location: PropTypes.string.isRequired,
+        open: PropTypes.bool.isRequired,
+        cleared: PropTypes.bool.isRequired,
+        calculateInputRect: PropTypes.func.isRequired,
+        onLoseVisibility: PropTypes.func.isRequired,
```
[the remark this answers is not in the document]
Comment (larkox) [reply components/suggestion/modal_suggestion_list.jsx]: (0/5) onHide seems like an active action. I prefer onLoseVisibility since it is more passive. The user does not actively hide the input, the input lose visibility when the user scrolls for whichever reason.

File: components/suggestion/suggestion_box.jsx (hunk 8 above)
Comment (larkox) [reply components/suggestion/suggestion_box.jsx]: Here, react-measure is using ResizeObserver to achieve their goal, but on our case we are mainly looking for changes in position, not so much in size. In order to check position, it exists IntersectionObserver, but it is only for checking the ration of intersection between two elements, which is not exactly what we need. I found ["solutions"](https://stackoverflow.com/questions/40251082/an-event-or-observer-for-changes-to-getboundingclientrect) to this using IntersectionObserver, but (1/5) they feel too convoluted for our solution.

The only event that seems to gives us this kind of reactiveness is scroll. But since we only want to check the scroll on the modal, it makes more sense to keep the registering and handling of the event on the "modal_suggestion_list", instead of on the "suggestion_box". Therefore, we should pass some way to calculate the elements size.

The other way to do this, is as your first suggestion: to move the calculateInputRect logic to modal_suggestion_list, and simply pass the input reference through the props. That way everything is handled in the same place.

File: components/suggestion/modal_suggestion_list.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/suggestion/modal_suggestion_list.jsx]: Yeah, I'd suggest the destructuring method if we were using those props, but since we aren't, I think `Reflect.deleteProperty` is the right way to go here

File: components/autocomplete_selector.jsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply components/autocomplete_selector.jsx]: `PropTypes.elementType` would probably be the correct type to use for things like `listComponent={MyComponent}`. `node` is for something that's already been rendered into jsx like if this were `listComponent={<MyComponent/>}`.

File: components/suggestion/modal_suggestion_list.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (ali-farooq0) [reply components/suggestion/modal_suggestion_list.jsx]: @larkox 
Ahh fair enough, how about this instead?


```
const {
	onLoseVisibility: _,
	calculateinputRect: _,
	...otherProps
} = this.props;
```

The reason i prefer this is because everything is inside one single block as opposed to the current approach of having two operations, one is destructing and then removing items.

File: components/suggestion/modal_suggestion_list.jsx [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+import SuggestionList from 'components/suggestion/suggestion_list.jsx';
+import {getClosestParent} from 'utils/utils.jsx';
+
+export default class ModalSuggestionList extends React.PureComponent {
+    static propTypes = {
+        location: PropTypes.string.isRequired,
+        open: PropTypes.bool.isRequired,
+        cleared: PropTypes.bool.isRequired,
+        inputRef: PropTypes.object.isRequired,
```
Comment (ali-farooq0) [inline components/suggestion/modal_suggestion_list.jsx]: 👍

File: components/suggestion/modal_suggestion_list.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (larkox) [reply components/suggestion/modal_suggestion_list.jsx]: That approach yields the following lint error:
`'_' is assigned a value but never used`

The current approach is being used in several parts of the code, and I would consider it a standard in the code right now. Nevertheless, if you come up with any other idea, I am more than happy to try.

[the remark this answers is not in the document]
Comment (ali-farooq0) [reply components/suggestion/modal_suggestion_list.jsx]: thanks  for giving it a try @larkox! :)
